### PR TITLE
Add a recomposer command to upgrade a composer schema.

### PR DESCRIPTION
**Pull request by maxime-rainville**
Added a command to upgrade your `composer.json` file.

# High level explanation of what's going on
You can use the `ComposerSchema` class to interact with a compose.json file. This class as an `upgrade` method that accepts an array of _rules_ that will be applied to the requirements in the file.

The main rule is called `rebuild`. This rule will create an empty composer file and try add back all the dependencies one-by-one but with a constraint on SS4.

Once it's done that, it will try to find a combination of recipe that could replace some of the packages.

At the end of the process, you should always have a composer file with all the dependencies you started of with (or equivalent recipes). Your dependencies might/probably will have conflicts that will need to be resolved manually.

# Patience is a virtue
This command can take quite bit of time to run, because it's making a lot of calls to composer and fetching a lot dependencies from Packagist. Running it on a composer file with 50 dependencies can take 6-10 min.

I've tried to add a bit of caching, but it's kind-of-dirty and hackish. Also because we're fetching things from the internet some of the unit test will have non-deterministic results (e.g.: Someone updating their package version online could break a test). I tried the write the test to minimize the risk of this happening, but short of completely mocking up composer, itm ight not be possible to work around this.

# Parent issue
#45

### CODE REVIEW

File: tests/Composer/ComposerExecTest.php [hunk 1]
```diff
@@ -208,10 +208,13 @@ public function testCacheDir()
...
-        $this->assertEquals(
+        $this->assertContains(
             $composer->getCacheDir(),
-            $_SERVER['HOME'] . '/.cache/composer',
-            'On unix system the composer cache is in ~/.cache/composer'
+            [
+                $_SERVER['HOME'] . '/.composer/cache',
+                $_SERVER['HOME'] . '/.cache/composer',
+            ],
+            'COmposer CacheDir is not in one of the expected location'
```
Comment (dhensby) [inline tests/Composer/ComposerExecTest.php]: Typo here

```diff
-'COmposer CacheDir is not in one of the expected location'
+'Composer CacheDir is not in one of the expected location'
```

File: src/Composer/ComposerExec.php [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+    protected $suppressErrors;
+
+    /**
+     * Instanciate a new ComposerExec.
+     * @param string  $workingDir Main Path where the composer project reside.
+     * @param string  $execPath Path to the composer executable.
+     * @param bool $suppressErrors Whatever errors should be supress. If false, composer errors will be printed on the
+     * CLI. Defaults to `true`. (This is meant for testing only.)
+     */
+    public function __construct(
```
Comment (unclecheese) [inline src/Composer/ComposerExec.php]: Minor point -- can you proofread this and tidy up the grammar?

File: src/Console/RecomposeCommand.php [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+        if ($write) {
+            $schema->setContents($change->newContents($schema->getFullPath()));
+            $output->writeln("Changes have been saved.");
+        } else {
+            $output->writeln("Changes not saved; Run with --write to commit to disk");
+        }
+
+
+        return null;
```
Comment (unclecheese) [inline src/Console/RecomposeCommand.php]: As it says in the docs, this function takes a _long_ time to run. Can we put some more output into it? A little bit of feedback would go a long way toward giving the user peace of mind that things are happening as intended.

File: src/Composer/Package.php [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+ */
+class Package
+{
+
+    /**
+     * List of supported Silverstripe Module.
+     * @var array
+     */
+    const SUPPORTED_MODULES = [
+        "assertchris/hash-compat",
```
[the remark this answers is not in the document]
Comment (robbieaverill) [reply src/Composer/Package.php]: Oh that’s what you meant! There is a supported-modules repo on this GitHub account with a JSON file containing these but it’s probably out f date. We could use that and update it?

File: src/Composer/Packagist.php [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+        }
+
+        // We haven't found our folder, let's bail.
+        return false;
+    }
+
+    /**
+     * Try to write the content of a request to a cache folder.
+     */
+    protected function writeToCache(string $vendor, string $packageName, array $data)
```
[the remark this answers is not in the document]
Comment (robbieaverill) [reply src/Composer/Packagist.php]: Not super familiar on context here but you can’t type hint scalars in PHP 5.6, which we support on SS4

[the remark this answers is not in the document]
Comment (unclecheese) [reply src/Composer/Packagist.php]: Yeah, I checked, and this library requires PHP >= 7

[the remark this answers is not in the document]
Comment (robbieaverill) [reply src/Composer/Packagist.php]: Cool, so long as that’s cool... :-D

File: src/Composer/ComposerFile.php [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+    {
+        // If we haven't been provided any content, assume we are validating the current file.
+        if ($content === null) {
+            $content = $this->composerJson;
+        }
+
+        // If we have an array, JSON code it
+        if (is_array($content)) {
+            // This is a bit of silly hack. If require is empty it will be outputted. as an array. `composer validate`
+            // expect it to be an object, so we just had a dummy key pair in there to make sure it always generated as
```
[the remark this answers is not in the document]
Comment (maxime-rainville) [reply src/Composer/ComposerFile.php]: I tried that, but it it also converts things that should be arrays. e.g: `["a", "b", "c"]` would become `{"1": "a", "2": "b", "3": "c"}`

File: src/Console/RecomposeCommand.php (hunk 3 above)
Comment (maxime-rainville) [reply src/Console/RecomposeCommand.php]: What about I update the output so it say something like this?
```
* Upgrading PHP constraint
* Rebuilding dependencies
  * Requiring PHP ... ✅
  * Requiring ext-json ...  ✅
  * Requiring silverstripe/recipe-core ... ✅
  * Requiring silverstripe/recipe-cms ... ✅
  * Requiring old-ss3/package-without-ss4 ... ❎
  * Requiring silverstripe/contentreview ... ✅
  * Requiring silverstripe/sharedraftcontent ... ✅
  * Requiring symbiote/silverstripe-advancedworkflow ...✅
  * Manually re-adding failed dependencies
* Switching to recipes
 * Switching to silverstripe/recipe-collaboration
   * Removing silverstripe/recipe-cms
   * Removing silverstripe/contentreview
   * Removing silverstripe/sharedraftcontent
   * Removing symbiote/silverstripe-advancedworkflow
* Using strict version constraint
* Showing difference
... diff here ...

Some of your dependencies could not be resolved. You'll need to manually correct those failed dependencies and run a `composer update`.
```

File: src/Composer/ComposerExec.php [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+        }
+
+        // Redirect STDERR to suppress errors.
+        if ($this->suppressErrors) {
+            $statement .= ' 2>&1';
+        }
+
+        // Execute our command.
+        $return = exec($statement, $output, $exitCode);
+
```
[the remark this answers is not in the document]
Comment (maxime-rainville) [reply src/Composer/ComposerExec.php]: Dammit. Wish I had use that instead. I'll work it back in ... it shouldn't be too much trouble and will save us work down the line.

File: src/Composer/ComposerFile.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (unclecheese) [reply src/Composer/ComposerFile.php]: Ah, true. OK, plan B.

```php
$content['require'] = (object) []
```

^ Works in my testing :)

[the remark this answers is not in the document]
Comment (maxime-rainville) [reply src/Composer/ComposerFile.php]: Cool. Didn't know you could do that. I've applied your suggested fix to the validate method and I've also added to the part that saves the build the JSON output.

**Comment by maxime-rainville** [comment]
Ready for round 2.

The console output is a lot more chatty now. I've also switch to using `SymfonyStyle` to make it extra pretty to look at. (I resisted the urge to add pretty emojii everywhere)

As suggested by @unclecheese , I've switched to using Symfony Process for calling composer. This allowed me to add a pretty dot animation to show an ongoing process.

After writing the changes to file, a `composer update` will automatically as well.

I've adjusted the linting rules to show errors when function comments are missing, however I've restricted it to my specific bits, otherwise I would have had to fix everything else.

I've also added an extra test for PHP 7.2 to future proof ourselves. I've tried adding a PHP7.0 test but it looks like this is well and truly broken.

**Pull request merged**